mantidproject/mantid
# Rebin2D changed to work with fractional area tracking on plain Workspace2D

**Description of work.**

Fixes #24269 

Changes `Rebin2D` so that it will use fractional area tracking for non-`RebinnedOutput` workspaces, with fractional area assumed to be unity. There was a condition imposed previously in a PR to correct the error propagation of `RebinnedOutput` workspaces which required a `RebinnedOutput` workspace in `Rebin2D` but this is too restrictive and leads to incorrect rebinning for Energy-2theta slices in MSlice, as reported in #24269.

**Report to:** [R.Bewley / D.T.Adroja]. 

**To test:**

Run the script:

```
ws1 = Load('MER42567_Ei80.00meV_rings183.nxspe')
ws1tth2D = ConvertSpectrumAxis(ws1, 'Theta', 'Direct', 80)
ws1q = SofQW3(ws1, '0,4,4', 'Direct', 80, EAxisBinning='-20,0.5,70')
ws1q = NormaliseSpectra(ws1q)
ws1tth = Rebin2D(ws1tth2D, '-20,0.5,70', '0,40,40', UseFractionalArea=True, Transpose=False)
ws1tth = NormaliseSpectra(ws1tth)
plotSpectrum([ws1q, ws1tth], 0)

ws2 = Load('LET55372_6.78meV_OneToOne.nxspe')
ws2tth2D = ConvertSpectrumAxis(ws2, 'Theta', 'Direct', 6.78)
ws2q = SofQW3(ws2, '1,0.5,1.5', 'Direct', 6.78, EAxisBinning='-0.5,0.02,1.5')
ws2q = NormaliseSpectra(ws2q)
ws2tth = Rebin2D(ws2tth2D, '-0.5,0.02,1.5', '40,40,80', UseFractionalArea=True, Transpose=False)
ws2tth = NormaliseSpectra(ws2tth)
plotSpectrum([ws2q, ws2tth], 0)
```

Or use MSlice as described in the issue and check that the cuts integrating over `|Q|` and over `2theta` agree. Additionally, run this script to check that `ComputeIncoherentDOS`, which runs `Rebin2D` as a child algorithm works ok (the two cuts, `dosq` and `dostth` should quite close to each other).

```
dostth = ComputeIncoherentDOS(ws1tth2D, Temperature=5, EnergyBinning='0,0.5,70', TwoThetaSumRange='65,130')
ws1q2D = SofQW3(ws1, '0,0.01,13', 'Direct')
dosq = ComputeIncoherentDOS(ws1q2D, Temperature=5, EnergyBinning='0,0.5,70', QSumRange='7,14')
plotSpectrum([dosq, dostth], 0)
```


Fixes #24269. 




---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
